src/lua/protobuf.lua

```diff
@@ -0,0 +1,608 @@
+        id_by_value = id_by_value,
+        value_by_id = value_by_id,
+    }
+end
+
+-- Create a protocol object that stores message
+-- data needed for encoding.
+--
+-- Accepts protocol definition using protobuf.message
+-- and protobuf.enum functions as in example.
+--
+-- protocol_def = {
+--     protocol.message(<message_name>, <message_def>),
+--     protocol.enum(<enumc_name>, <enum_def>),
+--     <...>
+-- }
+local function protocol(protocol_def)
+    local res = {}
+    -- Declaration table is used to check forward declarations
+    -- false -- this type used as the field type in the message was not defined
+    -- true -- this field type was defined
+    local declarations = {}
+    for _, def in pairs(protocol_def) do
+        if declarations[def.name] then
+            error(('Double definition of name %q'):format(def.name))
+        end
+        if def.type == 'message' then
+            for _, field_def in pairs(def.field_by_id) do
+                local standard = scalars[field_def.type] ~= nil
+                local declared = declarations[field_def.type]
+                if not standard and not declared then
+                    declarations[field_def.type] = false
+                end
+            end
+        end
+        declarations[def.name] = true
+        res[def.name] = def
+    end
+    -- Detects a message or a enum that is used as a field type in message
+    -- but not defined in protocol. Allows a type be defined after usage
+    for def_type, declared in pairs(declarations) do
+        if not declared then
+            error(('Type %q is not declared'):format(def_type))
+        end
+    end
+    return setmetatable(res, protocol_mt)
+end
+
+local function is_number64(value)
+    if type(value) == 'cdata' and (ffi.istype(int64_t, value) or
+        ffi.istype(uint64_t, value)) then
+        return true
+    end
+end
+
+local function validate_length(value)
+    local MAX_LEN = 2^32
+    if string.len(value) > MAX_LEN then
+        error("Too long string to be encoded")
+    end
+end
+
+local function encode_len(value, field_id)
+    local wire_type = 2
+    return string.format('%s%s%s',
+        wireformat.encode_uint(bit.lshift(field_id, 3) + wire_type),
+        wireformat.encode_uint(string.len(value)),
+        value)
+end
+
+local encode_field
+
+local encode
+
+local function encode_repeated(protocol, field_def, data)
+    local buf = {}
+    local scalar_def = scalars[field_def.type]
+    if type(data) ~= 'table' then
+        error('For repeated fields table data are needed')
+    elseif scalar_def ~= nil and scalar_def.encode_as_packed then
+        for _, value in pairs(data) do
+            table.insert(buf, string.sub(encode_field(protocol, field_def,
+                value, true), 2))
+        end
+        return encode_len(table.concat(buf), field_def.id)
+    else
+        for _, value in pairs(data) do
+            table.insert(buf, encode_field(protocol, field_def, value, true))
+        end
```

Comment:
First, `pairs()` traverses a table in an arbitrary order. There are some details [here](https://github.com/tarantool/crud/issues/235), but the key point is that Lua/LuaJIT have no guarantees regarding an order of values given by `pairs()`.

It means that `["a", "b", "c"]` may become `["b", "c", "a"]` after encoding and decoding. It looks very undesirable.

Next, the [RFC document](https://www.notion.so/tarantool/RFC-Protocol-Buffers-encoder-decoder-17ff2ff599974f9a9d0f39d9ab0a5284) requires the implementation to verify that a `repeated` value is a table with `number` indexes from 1 to N without any holes.

I would point my code that implements such a check.

https://github.com/tarantool/tarantool/blob/05d03a1c58d9fb5b46078fefbe2a3049df39d052/src/box/lua/config/utils/schema.lua#L429-L478

After this check we can just traverse the table using `ipairs()`, which guarantees the traversal order with increasing keys for an array-like table that starts from 1 and has no holes.

Reply:
Added check for arrays and covered it with tests